src/v/cloud_topics/level_zero/gc/level_zero_gc.cc

```diff
@@ -77,19 +77,183 @@ class epoch_source_cluster_impl : public level_zero_gc::epoch_source {
+                      maybe_max_gc_epoch.value());
+
+                    const auto p_id = partition_status.first;
+                    auto& partition_epochs = result[tp_ns];
+                    const auto it = partition_epochs.find(p_id);
+                    if (it == partition_epochs.end()) {
+                        partition_epochs.try_emplace(p_id, max_gc_epoch);
+                    } else {
+                        it->second = std::max(it->second, max_gc_epoch);
+                    }
+                }
+            }
+
+            /*
+             * A scheduling point is injected after looking at each node's
+             * report. We own the list of node reports which is a set of shared
+             * pointers, so iteration is safe, and the scheduling point is
+             * intended to help avoid reactor stalls. If we need to inject
+             * scheduling points at a finer granularity we'll need to take a
+             * closer look at concurrency rules of the reports themselves.
+             */
+            co_await seastar::maybe_yield();
+
+            if (as && as->abort_requested()) {
+                co_return std::unexpected("Abort requested");
+            }
+        }
+
+        co_return result;
+    }
+
     seastar::future<std::expected<std::optional<cluster_epoch>, std::string>>
-    max_gc_eligible_epoch(seastar::abort_source*) override {
+    max_gc_eligible_epoch(seastar::abort_source* as) override {
+        /*
+         * First retrieve a consistent snapshot of cloud topic partitions. This
+         * establishes a set of partitions from which we must obtain an epoch
+         * bound on garbage collection.
+         */
+        auto partitions = co_await get_partitions(as);
+        if (!partitions.has_value()) {
+            co_return std::unexpected(partitions.error());
+        }
+        if (partitions.value().partitions.empty()) {
+            co_return std::nullopt;
+        }
+
         /*
-         * TODO(noah): missing an integration of epochs with the partition
-         * table. for now we report no eligible epoch.
+         * Next we retrieve the latest reported epoch bounds from all cloud
+         * topic partitions. The source for this information is distributed,
+         * while the source for the `partitions` set above is centralized, and
+         * this is why we have these two different collection steps.
          */
-        co_return std::nullopt;
+        auto gc_epochs = co_await get_partitions_max_gc_epoch(as);
+        if (!gc_epochs.has_value()) {
+            co_return std::unexpected(gc_epochs.error());
+        }
+
+        /*
+         * The final result begins as the maximum epoch for the given snapshot.
+         * Below we merge the two result sets and walk the final result back to
+         * account for the partition with the smallest eligible gc epoch.
+         */
+        auto result = cluster_epoch(partitions.value().last_applied());
```

Comment:
I don't quite understand the relationship between last applied topic table revision and the partition epochs.

Reply:
Did you read through the "L0 data management" tab in the cloud topics RFC (and especially the appendix describing edge cases) and still have the same question?

I do need to figure out how to document the algorithm in the code more effectively.

Reply:
> Do I have that right?

Yep! We should get together and make a document or picture. I think the entire algorithm is pretty simple, but it's a lot of details to track and put together. For that reason I think it's kind of hard to understand in "code form".